### PR TITLE
feat: varint protocol version

### PR DESCRIPTION
**Pull request by QuantumExplorer**
## Issue being fixed or feature implemented
We used to using a u32 prepended to each message for the version. This PR changes this to a varint. This is to save space.

## Checklist:

- [X] I have performed a self-review of my own code
- [X] I have commented my code, particularly in hard-to-understand areas
- [X] I have added or updated relevant unit/integration/functional/e2e tests
- [ ] I have made corresponding changes to the documentation

**For repository code-owners and collaborators only**
- [ ] I have assigned this pull request to a milestone

### CODE REVIEW

**Comment by qrayven** [comment]
LGTM, I didn't look that much into details, but I'm just wondering whether decoding collisions are possible. I mean, when the continuation bytes of varInt match the bytes from the payload

**Review by qrayven — APPROVED** [review]

File: packages/rs-dpp/src/identity/identity.rs [hunk 1]
```diff
@@ -169,9 +170,11 @@ impl Identity {
     }
 
     pub fn from_cbor(identity_cbor: &[u8]) -> Result<Self, ProtocolError> {
-        let (protocol_version_bytes, identity_cbor_bytes) = identity_cbor.split_at(4);
-
-        let protocol_version = deserializer::get_protocol_version(protocol_version_bytes)?;
```
Comment (shumkov) [inline packages/rs-dpp/src/identity/identity.rs:174]: Use `decode_protocol_entity`

File: packages/rs-dpp/src/document/mod.rs [hunk 2]
```diff
@@ -158,17 +159,18 @@ impl Document {
     }
 
     pub fn from_buffer(cbor_bytes: impl AsRef<[u8]>) -> Result<Document, ProtocolError> {
-        let (protocol_version_bytes, document_cbor_bytes) = cbor_bytes.as_ref().split_at(4);
+        let (protocol_version, offset) =
```
Comment (shumkov) [inline packages/rs-dpp/src/document/mod.rs:162]: Use `decode_protocol_entity`

File: packages/rs-dpp/src/document/document_stub.rs [hunk 3]
```diff
@@ -262,9 +269,13 @@ impl DocumentStub {
...
         }
+        let (protocol_version, offset) =
+            u32::decode_var(document_cbor.as_ref()).ok_or(ProtocolError::DecodingError(
+                "contract cbor could not decode protocol version".to_string(),
+            ))?;
+        let (_, read_document_cbor) = document_cbor.as_ref().split_at(offset);
 
-        let (version, read_document_cbor) = document_cbor.split_at(4);
-        if !check_protocol_version_bytes(version) {
+        if !check_protocol_version(protocol_version) {
```
Comment (shumkov) [inline packages/rs-dpp/src/document/document_stub.rs:278]: Let's remove this function at all

File: packages/rs-dpp/src/document/document_stub.rs [hunk 4]
```diff
@@ -262,9 +269,13 @@ impl DocumentStub {
                 DataContractError::FieldRequirementUnmet("invalid document id"),
             ));
         }
+        let (protocol_version, offset) =
```
Comment (shumkov) [inline packages/rs-dpp/src/document/document_stub.rs:272]: Use `decode_protocol_entity`

File: packages/rs-dpp/src/document/document_stub.rs [hunk 5]
```diff
@@ -180,8 +181,12 @@ impl DocumentStub {
...
         owner_id: Option<[u8; 32]>,
     ) -> Result<Self, ProtocolError> {
-        let (version, read_document_cbor) = document_cbor.split_at(4);
-        if !check_protocol_version_bytes(version) {
+        let (protocol_version, offset) =
+            u32::decode_var(document_cbor.as_ref()).ok_or(ProtocolError::DecodingError(
+                "contract cbor could not decode protocol version".to_string(),
+            ))?;
+        let (_, read_document_cbor) = document_cbor.as_ref().split_at(offset);
+        if !check_protocol_version(protocol_version) {
```
Comment (shumkov) [inline packages/rs-dpp/src/document/document_stub.rs:189]: Let's remove this function at all. It does nothing

File: packages/rs-dpp/src/document/document_stub.rs [hunk 6]
```diff
@@ -180,8 +181,12 @@ impl DocumentStub {
         document_id: Option<[u8; 32]>,
         owner_id: Option<[u8; 32]>,
     ) -> Result<Self, ProtocolError> {
-        let (version, read_document_cbor) = document_cbor.split_at(4);
-        if !check_protocol_version_bytes(version) {
+        let (protocol_version, offset) =
```
Comment (shumkov) [inline packages/rs-dpp/src/document/document_stub.rs:184]: Use `decode_protocol_entity`

File: packages/rs-dpp/src/decode_protocol_entity_factory.rs [hunk 7]
```diff
@@ -14,8 +15,11 @@ impl DecodeProtocolEntity {
     pub fn decode_protocol_entity(
         buffer: impl AsRef<[u8]>,
     ) -> Result<(u32, JsonValue), ProtocolError> {
-        let (protocol_bytes, document_bytes): (&[u8], &[u8]) = buffer.as_ref().split_at(4);
-        let protocol_version = get_protocol_version(protocol_bytes)?;
+        let (protocol_version, offset) =
```
Comment (shumkov) [inline packages/rs-dpp/src/decode_protocol_entity_factory.rs:18]: Use `decode_protocol_entity`

File: packages/rs-dpp/src/data_contract/data_contract.rs [hunk 8]
```diff
@@ -134,12 +135,18 @@ impl DataContract {
     }
 
     pub fn from_cbor(cbor_bytes: impl AsRef<[u8]>) -> Result<Self, ProtocolError> {
-        let (protocol_version_bytes, contract_cbor_bytes) = cbor_bytes.as_ref().split_at(4);
-        let protocol_version = deserializer::get_protocol_version(protocol_version_bytes)?;
+        let (protocol_version, offset) =
```
Comment (shumkov) [inline packages/rs-dpp/src/data_contract/data_contract.rs:138]: Use `decode_protocol_entity`

**Review by shumkov — APPROVED** [review]

**Pull request merged**